trieste/acquisition/function/greedy.py

```diff
@@ -351,3 +355,127 @@ def __call__(self, x: TensorType) -> TensorType:
+            fantasized_data = {
+                tag: _generate_fantasized_data(model, pending_points, self._fantasize_method)
+                for tag, model in models.items()
+            }
+            models = {
+                tag: FantasizedGPRModel(model, fantasized_data[tag])
+                for tag, model in models.items()
+            }
+
+        return self._builder.prepare_acquisition_function(models, datasets)
+
+
+def _generate_fantasized_data(model, pending_points, fantasize_method):
+    if fantasize_method == "KB":
+        fantasized_obs, _ = model.predict(pending_points)
+    elif fantasize_method == "sample":
+        fantasized_obs = model.sample(pending_points, num_samples=1)  # TODO check dim
+    else:
+        raise NotImplementedError(
+            f"Fantasize method not supported. "
+            f"Expected KB or sample, received "
+            f"{fantasize_method}"
+        )
+    return Dataset(pending_points, fantasized_obs)
+
+
+class FantasizedGPRModel(ProbabilisticModel):
+    """
+    Creates new model from an existing GPR model and additional data.
+    This new model posterior is conditioned and the GPR data and the additional one.
+    """
+
+    def __init__(self, model: ProbabilisticModel, fantasized_data: Dataset):
+        """
+
+        :param model: a GPR model
+        :param fantasized_data: additional dataset to condition on
+        :raise NotImplementedError: If model is not of class GaussianProcessRegression.
+        """
+
+        if not isinstance(model, GaussianProcessRegression):
+            raise NotImplementedError(
+                f"FantasizedGPRModel only works with GPR models, received " f"{model.__repr__()}"
+            )
+
+        self._model = model
+        self._fantasized_data = fantasized_data
+
+    def predict(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        return self._model.conditional_predict_f(query_points, self._fantasized_data)
+
+    def predict_joint(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        return self._model.conditional_predict_joint(query_points, self._fantasized_data)
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        return self._model.conditional_predict_f_sample(
+            query_points, self._fantasized_data, num_samples
+        )
+
+    def predict_y(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        return self._model.conditional_predict_y(query_points, self._fantasized_data)
+
+    def get_observation_noise(self) -> TensorType:
+        return self._model.get_observation_noise()
+
+    def get_kernel(self) -> gpflow.kernels.Kernel:
+        return self._model.get_kernel()
+
+    def log(self) -> None:
+        return self._model.log()
+
+    def update(self, dataset: Dataset) -> None:
+        raise NotImplementedError
+
+    def optimize(self, dataset: Dataset) -> None:
+        raise NotImplementedError
```

Comment:
potentially we can do a super call on the model passed as an argument and that would instantiate all the original methods that we don't want to overwrite

Reply:
I actually don't want to pass those methods, because one shouldn't update or optimize an FantasizeModel

Reply:
then we overwrite those two methods, add 1 line with a super call and remove 6 unnecessary lines for the other 3 methods :)

Reply:
I'll stick to this version I think. I have removed the update and optimize methods though.